actors/verifreg/src/state.rs

```diff
@@ -1,51 +1,120 @@
 
         let empty_mapmap =
             MapMap::<_, (), Address, AllocationID>::new(store, HAMT_BIT_WIDTH, HAMT_BIT_WIDTH)
                 .flush()
-                .map_err(|e| anyhow::anyhow!("Failed to create empty multi map: {}", e))?;
+                .map_err(|e| {
+                    actor_error!(illegal_state, "Failed to create empty multi map: {}", e)
+                })?;
 
         Ok(State {
             root_key,
+            token,
             verifiers: empty_map,
-            verified_clients: empty_map,
             remove_data_cap_proposal_ids: empty_map,
             allocations: empty_mapmap,
             next_allocation_id: 1,
             claims: empty_mapmap,
         })
     }
+
+    // Adds a verifier and cap, overwriting any existing cap for that verifier.
+    pub fn put_verifier(
+        &mut self,
+        store: &impl Blockstore,
+        verifier: &Address,
+        cap: &DataCap,
+    ) -> Result<(), ActorError> {
+        let mut verifiers =
+            make_map_with_root_and_bitwidth::<_, BigIntDe>(&self.verifiers, store, HAMT_BIT_WIDTH)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load verifiers")?;
+        // .context("failed to load verifiers")?;
+        verifiers
+            .set(verifier.to_bytes().into(), BigIntDe(cap.clone()))
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to set verifier")?;
+        self.verifiers = verifiers
+            .flush()
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to flush verifiers")?;
+        Ok(())
+    }
+
+    pub fn remove_verifier(
+        &mut self,
+        store: &impl Blockstore,
+        verifier: &Address,
+    ) -> Result<(), ActorError> {
+        let mut verifiers =
+            make_map_with_root_and_bitwidth::<_, BigIntDe>(&self.verifiers, store, HAMT_BIT_WIDTH)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load verifiers")?;
+
+        verifiers
+            .delete(&verifier.to_bytes())
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to remove verifier")?
+            .context_code(ExitCode::USR_ILLEGAL_ARGUMENT, "verifier not found")?;
+
+        self.verifiers = verifiers
+            .flush()
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to flush verifiers")?;
+        Ok(())
+    }
+
+    pub fn get_verifier_cap(
+        &self,
+        store: &impl Blockstore,
+        verifier: &Address,
+    ) -> Result<Option<DataCap>, ActorError> {
+        let verifiers =
+            make_map_with_root_and_bitwidth::<_, BigIntDe>(&self.verifiers, store, HAMT_BIT_WIDTH)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load verifiers")?;
+        let allowance = verifiers
+            .get(&verifier.to_bytes())
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to get verifier")?;
+        Ok(allowance.map(|a| a.0.clone() as DataCap)) // TODO: can I avoid the clone?
```

Comment:
The HAMT always stores and returns a reference under the hood, have you tried changing the return value to an option &datacap?

Reply:
It's a good thought, I couldn't get the lifetimes to work. I'll commit this and we can try to round up the TODOs before merging to master.

Reply:
The HAMT always returns a reference and that reference's lifetime will be only as long as the life of `verifires` instance.
There might be some magical way to avoid it, but I don't think so. The compiler might be smart enough to notice that the original is getting destroyed right away, and it might avoid clone, although I would be surprised.